src/components/network-chart/NetworkChartRow.js

```diff
@@ -29,117 +30,109 @@ import type { NetworkPayload } from '../../types/markers';
   // a specific payload.
   +networkPayload: NetworkPayload | null,
-  +markerStyle: {
-    [key: string]: string | number,
-  },
+  +markerWidth: CssPixels,
+  +startPosition: CssPixels,
   +threadIndex: ThreadIndex,
 };
 
-type State = {
+type State = {|
   pageX: CssPixels,
   pageY: CssPixels,
   hovered: ?boolean,
-};
+|};
 
 export type NetworkChartRowBarProps = {
   +marker: Marker,
+  +markerWidth: CssPixels,
   // Pass the payload in as well, since our types can't express a Marker with
   // a specific payload.
   +networkPayload: NetworkPayload,
 };
 
-// This component splits a network marker duration in 4 different phases,
+// This component splits a network marker duration in different phases,
 // and renders each phase as a differently colored bar.
-// 1. request queue
-// 2. request
-// 3. response
-// 4. response queue (This is not calculated. We assume it is the rest of the duration.)
 const NetworkChartRowBar = (props: NetworkChartRowBarProps) => {
-  const { marker, networkPayload } = props;
+  const { marker, networkPayload, markerWidth } = props;
   const { start, dur } = marker;
 
-  // A marker does not always contain the same set of networkPayload on the start of
-  // the connection.
-  const queueStart =
-    networkPayload.secureConnectionStart ||
-    networkPayload.tcpConnectEnd ||
-    networkPayload.connectStart ||
-    networkPayload.domainLookupStart ||
-    0;
-
-  // The default for width of the phases is always zero.
-  let requestQueue: number = 0;
-  let request: number = 0;
-  let response: number = 0;
-
-  // Timestamps from network markers aren't adjusted when processing
-  // the gecko profile format. This is the reason why the start timestamp of a
-  // marker can be after the start timestamp of the request.
-  // To prevent false networkPayload visualization, we need this workaround.
-  // See https://github.com/firefox-devtools/profiler/issues/1493 for more detail.
-  if (
-    networkPayload.requestStart &&
-    start < networkPayload.requestStart &&
-    networkPayload.responseStart &&
-    start < networkPayload.responseStart
-  ) {
-    request =
-      (networkPayload.responseStart - networkPayload.requestStart) / dur * 100;
-  }
-
-  if (
-    networkPayload.responseEnd &&
-    start < networkPayload.responseEnd &&
-    networkPayload.responseStart &&
-    start < networkPayload.responseStart
-  ) {
-    response =
-      (networkPayload.responseEnd - networkPayload.responseStart) / dur * 100;
-  }
-
-  if (queueStart && start < queueStart && queueStart > 0) {
-    requestQueue = (queueStart - start) / dur * 100;
-  }
+  const barParts = [];
+  let previousValue = start;
+  let previousName = 'startTime';
 
-  // When we keep the default values (=zero), the
-  // response takes the full width of the bar to be more visible.
-  if (requestQueue + request + response === 0) {
-    response = 100;
-  }
+  PROPERTIES_IN_ORDER.filter(
```

Comment:
Could you add a small comment here describing the operation, since this `forEach` is working by mutation?

Reply:
Would you prefer that I use a map ?

Reply:
I tried using a `map` function but I felt the code was less readable (because I'd need to put the object in a temporary var). I'll add a comment.